File: packages/go/hushspec/regex_safety_test.go
package hushspec

import (
	"os"
	"path/filepath"
	"testing"
)

func TestValidRegexPatternsPass(t *testing.T) {
	spec := &HushSpec{
		HushSpecVersion: "0.1.0",
		Rules: &Rules{
			SecretPatterns: &SecretPatternsRule{
				Enabled: true,
				Patterns: []SecretPattern{
					{Name: "aws_key", Pattern: "AKIA[0-9A-Z]{16}", Severity: SeverityCritical},
					{Name: "private_key", Pattern: "-----BEGIN\\s+(RSA\\s+)?PRIVATE\\s+KEY-----", Severity: SeverityCritical},
				},
			},
			ShellCommands: &ShellCommandsRule{
				Enabled:           true,
				ForbiddenPatterns: []string{"(?i)rm\\s+-rf\\s+/", "curl.*\\|.*bash"},
			},
			PatchIntegrity: &PatchIntegrityRule{
				Enabled:           true,
				MaxImbalanceRatio: floatPtr(10.0),
				ForbiddenPatterns: []string{"(?i)disable[\\s_\\-]?(security|auth|ssl|tls)", "(?i)chmod\\s+777"},
			},
		},
	}
	result := Validate(spec)
	if !result.IsValid() {
		t.Fatalf("expected valid RE2 patterns to pass validation, got errors: %+v", result.Errors)
	}
}

func TestInvalidRegexSyntaxSecretPatterns(t *testing.T) {
	spec := &HushSpec{
		HushSpecVersion: "0.1.0",
		Rules: &Rules{
			SecretPatterns: &SecretPatternsRule{
				Enabled: true,
				Patterns: []SecretPattern{
					{Name: "bad", Pattern: "[unterminated", Severity: SeverityCritical},
				},
			},
		},
	}
	result := Validate(spec)
	if result.IsValid() {
		t.Fatal("expected invalid regex syntax to fail validation")
	}
	found := false
	for _, err := range result.Errors {
		if err.Code == "INVALID_REGEX" {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("expected INVALID_REGEX error code, got: %+v", result.Errors)
	}
}

func TestInvalidRegexSyntaxShellCommands(t *testing.T) {
	spec := &HushSpec{
		HushSpecVersion: "0.1.0",
		Rules: &Rules{
			ShellCommands: &ShellCommandsRule{
				Enabled:           true,
				ForbiddenPatterns: []string{"[invalid"},
			},
		},
	}
	result := Validate(spec)
	if result.IsValid() {
		t.Fatal("expected invalid regex in shell_commands to fail validation")
	}
}

func TestInvalidRegexSyntaxPatchIntegrity(t *testing.T) {
	spec := &HushSpec{
		HushSpecVersion: "0.1.0",
		Rules: &Rules{
			PatchIntegrity: &PatchIntegrityRule{
				Enabled:           true,
				MaxImbalanceRatio: floatPtr(10.0),
				ForbiddenPatterns: []string{"valid_pattern", "(unclosed"},
			},
		},
	}
	result := Validate(spec)
	if result.IsValid() {
		t.Fatal("expected invalid regex in patch_integrity to fail validation")
	}
}

func TestGoRegexpRejectsBackreference(t *testing.T) {
	spec := &HushSpec{
		HushSpecVersion: "0.1.0",
		Rules: &Rules{
			SecretPatterns: &SecretPatternsRule{
				Enabled: true,
				Patterns: []SecretPattern{
					{Name: "backref", Pattern: "(a)\\1", Severity: SeverityCritical},
				},
			},
		},
	}
	result := Validate(spec)
	if result.IsValid() {
		t.Fatal("expected backreference pattern to be rejected by Go regexp (RE2)")
	}
}

func TestGoRegexpRejectsLookahead(t *testing.T) {
	spec := &HushSpec{
		HushSpecVersion: "0.1.0",
		Rules: &Rules{
			ShellCommands: &ShellCommandsRule{
				Enabled:           true,
				ForbiddenPatterns: []string{"(?=foo)bar"},
			},
		},
	}
	result := Validate(spec)
	if result.IsValid() {
		t.Fatal("expected lookahead pattern to be rejected by Go regexp (RE2)")
	}
}

func TestGoRegexpRejectsLookbehind(t *testing.T) {
	spec := &HushSpec{
		HushSpecVersion: "0.1.0",
		Rules: &Rules{
			PatchIntegrity: &PatchIntegrityRule{
				Enabled:           true,
				MaxImbalanceRatio: floatPtr(10.0),
				ForbiddenPatterns: []string{"(?<=password:)\\s*\\S+"},
			},
		},
	}
	result := Validate(spec)
	if result.IsValid() {
		t.Fatal("expected lookbehind pattern to be rejected by Go regexp (RE2)")
	}
}

func TestBuiltInRulesetsPassValidation(t *testing.T) {
	rulesetDir := filepath.Join("..", "..", "..", "rulesets")
	rulesetFiles := []string{
		"default.yaml",
		"strict.yaml",
		"permissive.yaml",
		"ai-agent.yaml",
		"cicd.yaml",
		"remote-desktop.yaml",
	}

	for _, filename := range rulesetFiles {
		t.Run(filename, func(t *testing.T) {
			path := filepath.Join(rulesetDir, filename)
			data, err := os.ReadFile(path)
			if err != nil {
				t.Fatalf("failed to read ruleset %s: %v", filename, err)
			}
			spec, parseErr := Parse(string(data))
			if parseErr != nil {
				t.Fatalf("ruleset %s failed to parse: %v", filename, parseErr)
			}
			result := Validate(spec)
			if !result.IsValid() {
				t.Fatalf("ruleset %s failed validation: %+v", filename, result.Errors)
			}
		})
	}
}

func floatPtr(f float64) *float64 {
	return &f
}

File: packages/hushspec/tests/regex-safety.test.ts
import { describe, it, expect } from 'vitest';
import { parse } from '../src/parse.js';
import { validate, isSafeRegex } from '../src/validate.js';
import { parseOrThrow } from '../src/parse.js';

// ---------------------------------------------------------------------------
// isSafeRegex unit tests
// ---------------------------------------------------------------------------

describe('isSafeRegex', () => {
  it('accepts simple character class patterns', () => {
    expect(isSafeRegex('AKIA[0-9A-Z]{16}')).toBe(true);
  });

  it('accepts case-insensitive flag', () => {
    // (?i) is RE2-compatible but not valid JS RegExp; isSafeRegex only checks for non-RE2 features
    expect(isSafeRegex('(?i)disable[\\s_\\-]?(security|auth)')).toBe(true);
  });

  it('accepts dot-star with literal separator', () => {
    expect(isSafeRegex('curl.*\\|.*bash')).toBe(true);
  });

  it('accepts non-capturing groups', () => {
    expect(isSafeRegex('(?:key|token)\\s*[:=]\\s*[A-Za-z0-9]{32,}')).toBe(true);
  });

  it('accepts named groups (Python-style)', () => {
    expect(isSafeRegex('(?P<name>[a-z]+)')).toBe(true);
  });

  it('accepts anchors and word boundaries', () => {
    expect(isSafeRegex('^\\bfoo\\b$')).toBe(true);
  });

  it('accepts \\0 (null character, not a backreference)', () => {
    expect(isSafeRegex('\\0')).toBe(true);
  });

  it('rejects backreferences (\\1)', () => {
    expect(isSafeRegex('(a)\\1')).toBe(false);
  });

  it('rejects backreferences (\\2)', () => {
    expect(isSafeRegex('(a)(b)\\2')).toBe(false);
  });

  it('rejects named backreferences (\\k<name>)', () => {
    expect(isSafeRegex('(?<word>\\w+)\\k<word>')).toBe(false);
  });

  it('rejects positive lookahead (?=...)', () => {
    expect(isSafeRegex('foo(?=bar)')).toBe(false);
  });

  it('rejects negative lookahead (?!...)', () => {
    expect(isSafeRegex('foo(?!bar)')).toBe(false);
  });

  it('rejects positive lookbehind (?<=...)', () => {
    expect(isSafeRegex('(?<=password:)\\s*\\S+')).toBe(false);
  });

  it('rejects negative lookbehind (?<!...)', () => {
    expect(isSafeRegex('(?<!\\d)\\d{3}')).toBe(false);
  });

  it('rejects atomic groups (?>...)', () => {
    expect(isSafeRegex('(?>abc)')).toBe(false);
  });

  it('rejects possessive quantifier *+', () => {
    expect(isSafeRegex('a*+')).toBe(false);
  });

  it('rejects possessive quantifier ++', () => {
    expect(isSafeRegex('a++')).toBe(false);
  });

  it('rejects possessive quantifier ?+', () => {
    expect(isSafeRegex('a?+')).toBe(false);
  });

  it('rejects conditional patterns', () => {
    expect(isSafeRegex('(?(1)yes|no)')).toBe(false);
  });

  it('rejects named backreference (?P=name)', () => {
    expect(isSafeRegex('(?P<word>\\w+)(?P=word)')).toBe(false);
  });

  it('rejects subroutine calls \\g<name>', () => {
    expect(isSafeRegex('\\g<name>')).toBe(false);
  });
});

// ---------------------------------------------------------------------------
// Regex validation in parse/validate pipeline
// ---------------------------------------------------------------------------

describe('regex safety in validation', () => {
  it('accepts valid RE2-compatible secret pattern', () => {
    const result = parse(`
hushspec: "0.1.0"
rules:
  secret_patterns:
    patterns:
      - name: aws_key
        pattern: "AKIA[0-9A-Z]{16}"
        severity: critical
`);
    expect(result.ok).toBe(true);
  });

  it('rejects syntactically invalid regex', () => {
    const result = parse(`
hushspec: "0.1.0"
rules:
  secret_patterns:
    patterns:
      - name: bad
        pattern: "["
        severity: critical
`);
    expect(result.ok).toBe(false);
    if (!result.ok) {
      expect(result.error).toContain('valid regular expression');
    }
  });

  it('rejects regex with backreference in secret_patterns', () => {
    const result = parse(`
hushspec: "0.1.0"
rules:
  secret_patterns:
    patterns:
      - name: backref
        pattern: "(a)\\\\1"
        severity: critical
`);
    expect(result.ok).toBe(false);
    if (!result.ok) {
      expect(result.error).toContain('RE2');
    }
  });

  it('rejects regex with lookahead in shell_commands.forbidden_patterns', () => {
    const result = parse(`
hushspec: "0.1.0"
rules:
  shell_commands:
    forbidden_patterns:
      - "(?=foo)bar"
`);
    expect(result.ok).toBe(false);
    if (!result.ok) {
      expect(result.error).toContain('RE2');
    }
  });

  it('rejects regex with lookbehind in patch_integrity.forbidden_patterns', () => {
    const result = parse(`
hushspec: "0.1.0"
rules:
  patch_integrity:
    max_imbalance_ratio: 10.0
    forbidden_patterns:
      - "(?<=password:)\\\\s*\\\\S+"
`);
    expect(result.ok).toBe(false);
    if (!result.ok) {
      expect(result.error).toContain('RE2');
    }
  });

  it('accepts valid JS-compatible patterns across all regex fields', () => {
    const result = parse(`
hushspec: "0.1.0"
rules:
  secret_patterns:
    patterns:
      - name: aws_key
        pattern: "AKIA[0-9A-Z]{16}"
        severity: critical
      - name: github_token
        pattern: "gh[ps]_[A-Za-z0-9]{36}"
        severity: critical
  shell_commands:
    forbidden_patterns:
      - "[Rr][Mm]\\\\s+-[Rr][Ff]\\\\s+/"
  patch_integrity:
    max_imbalance_ratio: 10.0
    forbidden_patterns:
      - "[Cc][Hh][Mm][Oo][Dd]\\\\s+777"
`);
    expect(result.ok).toBe(true);
  });

  it('rejects non-RE2 patterns even if they are valid JS', () => {
    const result = parse(`
hushspec: "0.1.0"
rules:
  secret_patterns:
    patterns:
      - name: lookahead
        pattern: "(?=secret)\\\\w+"
        severity: critical
`);
    expect(result.ok).toBe(false);
    if (!result.ok) {
      expect(result.error).toContain('RE2');
    }
  });
});

// ---------------------------------------------------------------------------
// Built-in rulesets: verify regex patterns are RE2-safe
// ---------------------------------------------------------------------------

describe('built-in ruleset patterns are RE2-safe', () => {
  const rulesetPatterns: Array<{ name: string; patterns: string[] }> = [
    {
      name: 'default.yaml',
      patterns: [
        'AKIA[0-9A-Z]{16}',
        'gh[ps]_[A-Za-z0-9]{36}',
        'sk-[A-Za-z0-9]{48}',
        '-----BEGIN\\s+(RSA\\s+)?PRIVATE\\s+KEY-----',
        '(?i)disable[\\s_\\-]?(security|auth|ssl|tls)',
        '(?i)skip[\\s_\\-]?(verify|validation|check)',
        '(?i)rm\\s+-rf\\s+/',
        '(?i)chmod\\s+777',
      ],
    },
    {
      name: 'strict.yaml',
      patterns: [
        'AKIA[0-9A-Z]{16}',
        'gh[ps]_[A-Za-z0-9]{36}',
        'sk-[A-Za-z0-9]{48}',
        'sk-ant-[A-Za-z0-9\\-]{95}',
        '-----BEGIN\\s+(RSA\\s+)?PRIVATE\\s+KEY-----',
        'npm_[A-Za-z0-9]{36}',
        'xox[baprs]-[0-9]{10,13}-[0-9]{10,13}[a-zA-Z0-9-]*',
        '(?i)(api[_\\-]?key|apikey)\\s*[:=]\\s*[A-Za-z0-9]{32,}',
        '(?i)disable[\\s_\\-]?(security|auth|ssl|tls)',
        '(?i)skip[\\s_\\-]?(verify|validation|check)',
        '(?i)rm\\s+-rf\\s+/',
        '(?i)chmod\\s+777',
        '(?i)eval\\s*\\(',
        '(?i)exec\\s*\\(',
        '(?i)reverse[_\\-]?shell',
        '(?i)bind[_\\-]?shell',
      ],
    },
    {
      name: 'ai-agent.yaml',
      patterns: [
        'AKIA[0-9A-Z]{16}',
        'gh[ps]_[A-Za-z0-9]{36}',
        'sk-[A-Za-z0-9]{48}',
        'sk-ant-[A-Za-z0-9\\-]{95}',
        '-----BEGIN\\s+(RSA\\s+)?PRIVATE\\s+KEY-----',
        '(?i)rm\\s+-rf\\s+/',
        '(?i)chmod\\s+777',
        'curl.*\\|.*bash',
        'wget.*\\|.*bash',
      ],
    },
    {
      name: 'cicd.yaml',
      patterns: [
        'AKIA[0-9A-Z]{16}',
        'gh[ps]_[A-Za-z0-9]{36}',
        '-----BEGIN\\s+(RSA\\s+)?PRIVATE\\s+KEY-----',
      ],
    },
  ];

  for (const { name, patterns } of rulesetPatterns) {
    it(`${name}: all patterns are RE2-safe`, () => {
      for (const pattern of patterns) {
        expect(isSafeRegex(pattern)).toBe(true);
      }
    });
  }

  // permissive.yaml and remote-desktop.yaml have no regex patterns to validate.
});
